### DISCUSSION TITLE
Tests organization, selection and types

### DISCUSSION THREAD

**Discussion by AljenU** [automaticanalysis/automaticanalysis#282] [General]
Related to #272 

Probably best to try to have different threads here for different subjects.

Currently, there are 4 long running tests to check whether AA is running correctly. These are now triggered as part of a CI workflow each time a new addition is (requested to be) merged into master. I think it is good to keep these, but to also extend with more tests.

The current tests are integration tests, that run a complete workflow, which is also why they take quite a while to run. Preferably there are also many small tests, that check specific functions. This to avoid that bugs that have been found (and that are not necessarily triggered in the integration tests) are reintroduced. For instance, the bug from #267 is not triggered by the integration tests.

Also, the Windows compatibility that is being worked on, should be captured in tests, so changes that are made later on do not again break the Windows compatibility.

Since it may be that certain tests will only run on Linux, and others only on Windows, there should be an easy way to select which tests to run when (although as many tests as possible should run on all platforms).

So as not to reinvent the wheel, i'd like to use the matlab testing framework to help with test selection (and other test related tasks). Various selectors are available there already, and ways to have shared setup functionality, and for later also utilities to get more information on code coverage.

1. Do you agree with the above analysis?
2. On aa_test.m: One question is how to merge the use of the matlab testing framework with the currently existing aa_test function. A similar function, for selecting what tests to run based on string inputs, would also be needed when using the matlab testing framework. It would have some additional selection options, but we can also keep the current ones. I would prefer to wrap the current tests also with the matlab testing framework, and further adapt aa_test to mainly be the test selection function. The extra code needed to correctly run the integration tests would go in the shared wrapper for those tests.
3. The framework's tag functionality can be used to distinguish tests. For long running tests, that should have a dedicated job in the CI file (all small tests can go together in 1 job), the tag could be "End-to-end", or "Integration". What would be a clear tag name?

Edit: A Matlab blog on tagging tests: https://blogs.mathworks.com/developer/2015/09/08/test-tags/

**Comment by tiborauer** [comment]
Ad 1.: We consider these tests actually short :), but I agree with your analysis.
Ad 2.: I am not familiar with the Matlab testing framework, but I am eager to learn more. :) I'd leave this question to @jones-michael-s, because aa_test is his creation. 
Ad 3.: Although I can see why "Integration" seems to be a fitting label, I think the current tests are better labelled as "Use case". Some tests integrate more than one tool, and the scripts are also named after use cases.

**Comment by jones-michael-s** [comment]
One way to make the testing quicker would be to use a single "omnibus" test rather than having a larger collection of scripts that test narrow functionality. I chose the latter philosophy when writing the new aa_test  thinking it would be good when changing specific parts of the code or when adding new modules. But running *all* of these scripts is inefficient because they share modules (preprocessing, for example, which is the most time consuming processing in some tests).

Perhaps both types of scripts are needed? For example, if I modify aamod_waveletdespike, I don't really need to test dartel or AROMA. But if I change something in the engine, then much wider testing is needed.

**Comment by tiborauer** [comment]
I do not think @AljenU suggests "omnibus" test. Besides, all tests run in parallel on CI, so a single "omnibus" test would take longer (and would not fit in the hardware limitations - time and diskspace). I think @AljenU suggest "atomising" the test further, but I do not think it is possible for **use-cases**.
